rfd/0067-database-access-aws-redis

```diff
@@ -0,0 +1,221 @@
+    // Key is the key path of the secret.
+    Key string
+    // Value is the value of the secret.
+    Value string
+    // Version is the version of the secret value.
+    Version string
+    // Created is the creation time of this version. Can be used to calculate
+    // if a secret value is expired.
+    Created time.Time
+}
+
+// CreateRequest defines the request to create a new secret.
+type CreateRequest struct {
+    // Key is the key path of the secret.
+    Key string
+    // Value is the value of the secret to set. A random value is generated if
+    // empty.
+    Value string
+    // RandomValueLength is the length of the random value to generate.
+    RandomValueLength int
+}
+
+// PutValueRequest defines the request to create a new secret version.
+type PutValueRequest struct {
+    // CurrentValue is the current secret value. Key in CurrentValue is used to
+    // identify the secret. Other attributes can be used for test-and-set
+    // operations.
+    CurrentValue *SecretValue
+    // Value is the value of the secret to set. A random value is generated if
+    // empty.
+    Value string
+    // RandomValueLength is the length of the random value to generate.
+    RandomValueLength int
+}
+
+// Secrets defines an interface for managing secrets. A secret consists of a
+// key path and a list of versions which hold copies of current or past secret
+// values.
+type Secrets interface {
+    // Create creates a new secret.
+    Create(req CreateRequest) (*SecretValue, error)
+
+    // Delete deletes the secret for provided path.
+    Delete(key string) error
+
+    // PutValue creates a new secret version for the secret. Implementations
+    // must guarantee that only the first of many calls to PutValue succeeds,
+    // when req.CurrentValue are the same.
+    PutValue(req PutValueRequest) (*SecretValue, error)
+
+    // GetValue returns the secret value for provided version. Besides version
+    // string returned from PutValue, two specials versions "CURRENT" and
+    // "PREVIOUS" can also be used to retrieve the current and previous
+    // versions respectively. If version is empty, "CURRENT" is used.
+    GetValue(key string, version string) (*SecretValue, error)
+}
+```
+It is up to the caller to call `PutValue` periodically to rotate the secret.
+
+#### Secret store - Random password generation
+[sethvargo/go-password](https://github.com/sethvargo/go-password) can be used to generate random passwords. <sup>[3]<sup>
+
+A minimum length of 16 should be enforced for generating random passwords. A mix of lower case letters, upper case letters, digits, and symobols should be also enforced.
+
+#### Secret store - AWS Secrets Manager
+With `iam:PutUserPolicy` and `iam:PutRolePolicy` permissions, Teleport service can grant required permissions to itself to perform Secrets Manager APIs. Users can limit the permissions by setting the following permissions boundaries (preferably using bootstrap configurator):
```

Comment:
At which point will it attempt to acquire these permissions?

Reply:
As discussed, users have to grant permission from their side and configurator can be used to help generate the policy. updated rfd accordingly